metrics/metrics.go

```diff
@@ -2,74 +2,105 @@ package metrics
 
 import (
 	"net/http"
+	"sync"
 	"time"
 
 	"go.uber.org/zap"
 
+	prom "github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+
 	"github.com/ethereum/go-ethereum/metrics"
 	gethprom "github.com/ethereum/go-ethereum/metrics/prometheus"
 	"github.com/status-im/status-go/logutils"
 
-	prom "github.com/prometheus/client_golang/prometheus"
-	"github.com/prometheus/client_golang/prometheus/promhttp"
-
 	"github.com/status-im/status-go/common"
 )
 
 // Server runs and controls a HTTP pprof interface.
 type Server struct {
-	server *http.Server
+	server        *http.Server
+	handlers      map[string]http.Handler
+	handlersMutex sync.RWMutex
 }
 
 func NewMetricsServer(address string, r metrics.Registry) *Server {
 	mux := http.NewServeMux()
-	mux.Handle("/health", healthHandler())
-	mux.Handle("/metrics", Handler(r))
-	p := Server{
+
+	s := &Server{
+		handlers: make(map[string]http.Handler),
 		server: &http.Server{
 			Addr:              address,
 			ReadHeaderTimeout: 5 * time.Second,
 			Handler:           mux,
 		},
 	}
-	return &p
+
+	// we disable compression because geth doesn't support it
+	opts := promhttp.HandlerOpts{DisableCompression: true}
+	// register status handler
+	s.RegisterHandler("status", promhttp.HandlerFor(prom.DefaultGatherer, opts))
+
+	// register geth handler
+	if r != nil {
+		s.RegisterHandler("geth", gethprom.Handler(r))
+	}
+
+	mux.Handle("/health", healthHandler())
+	mux.Handle("/metrics", s.metricsHandler())
+
+	return s
 }
 
-func healthHandler() http.Handler {
+// RegisterHandler adds a new metrics provider with a given name
+func (s *Server) RegisterHandler(name string, handler http.Handler) {
+	s.handlersMutex.Lock()
+	defer s.handlersMutex.Unlock()
+	s.handlers[name] = handler
+}
+
+// UnregisterHandler removes a metrics provider
+func (s *Server) UnregisterHandler(name string) {
+	s.handlersMutex.Lock()
+	defer s.handlersMutex.Unlock()
+	delete(s.handlers, name)
+}
+
+// metricsHandler creates the combined metrics handler
+func (s *Server) metricsHandler() http.Handler {
+
 	return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
-		_, err := w.Write([]byte("OK"))
-		if err != nil {
-			logutils.ZapLogger().Error("health handler error", zap.Error(err))
+
+		// Write all dynamic metrics
+		s.handlersMutex.RLock()
+		for _, handler := range s.handlers {
+			if handler != nil {
```

Comment:
Let's just check the handler when adding it. And assume no handlers are `nil` here.
Because thi condition will be executed on each `/metrics` call, while the `RegisterHandler` is called only a few times.